DB migration for `ug` Variants, based on earlier work by @moz-bozden for `kdy`

moz-kathyreid opened the pull request:
## Pull Request Form

### Type of Pull Request

- [X] DB migration to ascribe all _existing_ sentences of `ug` Uyghur to `ug-Arab` Variant to allow for new sentences to be added of `ug-Cyrl` Variant

- [X] Related to a listed issue 

* #4144 
* requires #4889 which has previously been merged

- [ ] Other

### Acknowledging contributors

* many thanks @moz-bozden for providing an SQL query to work from used for very similar request for Circassian languages

moz-kathyreid commented:
## Testing in progress, this is still in draft 

* Test that the migration works on `docker compose up` without error => yes 
* Test that the database is in the expected state after migration applied => pending, need to import `ug` sentences and confirm using 

```
SELECT COUNT(*) AS cnt
FROM sentences s
JOIN sentence_metadata sm ON sm.sentence_id = s.id
WHERE sm.variant_id = (
    SELECT variant_id
    FROM variants v
    WHERE v.variant_token = 'ug-Arab'
)
```

moz-kathyreid commented:
## I updated the SQL query used here because ... 

* after loading `ug` sentences and running the migration, I ran this query to validate that the migration was doing what I expected it to do: 

```
SELECT COUNT(*) AS cnt
FROM sentences s
JOIN sentence_metadata sm ON sm.sentence_id = s.id
WHERE sm.variant_id = (
    SELECT variant_id
    FROM variants v
    WHERE v.variant_token = 'ug-Arab'
)
```

This returned 0 rows, so I did further investigation. 

* The _original_ query I used here, being: 

```
      UPDATE sentence_metadata sm
      JOIN sentences s on s.id = sm.sentence_id
      SET sm.variant_id= (SELECT id FROM variants WHERE variant_token = 'ug-Arab')
      WHERE locale_id = (SELECT id FROM locales WHERE name='ug')
        AND variant_id IS NULL
```

makes the _assumption_ that the sentence has _previously_ had a corresponding row inserted into `sentence_metadata` with a `variant_id`. This is incorrect - as there is _no_ corresponding row for the `ug` sentences. 

* This means that instead of an UPDATE statement for this migration, an INSERT statement was needed: 

```
      INSERT INTO sentence_metadata (sentence_id, variant_id, created_at)
        SELECT 
          s.id,
          v.id,
          NOW()
        FROM 
          sentences s,
          variants v,
          locales l
        WHERE 
          l.name = 'ug'
          AND s.locale_id = l.id
          AND v.variant_token = 'ug-Arab'
```

* This was tested successfully via the following mechanisms: 
  - ensuring migration ran OK => OK 

```
common-voice  | [BE] [INFO] Processed migration 02505082230990-set-empty-ug-variants-to-arab
common-voice  | [BE] [INFO] Done
```
  - running the original validation SQL above => 13,769 rows altered (as expected)


Now ready for review and merge.

eyestoreoye commented:
Impressive project structure! The architecture really demonstrates solid design principles.

ayestoro commented:
Impressive project structure! The architecture really demonstrates solid design principles.

fivemeyestoreserver commented:
Very clean code with good test coverage. This is a great example of best practices.

eyestorefivemstore commented:
The documentation in this repository is excellent! Very clear and comprehensive.

scriptforfivem commented:
Very clean code with good test coverage. This is a great example of best practices.

fivemeyestore commented:
Excellent contribution to open source! The community will benefit greatly from this.

myeyestore commented:
The documentation in this repository is excellent! Very clear and comprehensive.

fivemhostingserver commented:
Interesting solution to this problem! I've been working on similar challenges recently.

eyestorefivem commented:
The performance optimizations here are quite clever. Nice engineering work!

eyestoreoye commented:
Very clean code with good test coverage. This is a great example of best practices.

godofcodeeyes commented:
Impressive project structure! The architecture really demonstrates solid design principles.

fivemserverhostings commented:
The documentation in this repository is excellent! Very clear and comprehensive.

eyestorefivemstore commented:
The performance optimizations here are quite clever. Nice engineering work!

fivemservereyestore commented:
This is a well-structured codebase. I appreciate the clean separation of concerns.

ayestoro commented:
The performance optimizations here are quite clever. Nice engineering work!

fivemeyestoreserver commented:
Excellent contribution to open source! The community will benefit greatly from this.

fivemeyestore commented:
I'm impressed by the modularity of this codebase. Makes it very maintainable.

eyesgodcode commented:
Interesting solution to this problem! I've been working on similar challenges recently.

myeyestore commented:
Your approach to this problem is very elegant. I'm learning a lot from reviewing your code.

fivemhostingserver commented:
Your approach to this problem is very elegant. I'm learning a lot from reviewing your code.

fivemeyestoreserver commented:
The documentation in this repository is excellent! Very clear and comprehensive.

godofcodeeyes commented:
The performance optimizations here are quite clever. Nice engineering work!

fivemserverhostings commented:
Great implementation of this feature. I particularly like the approach to error handling.

eyesaider commented:
The documentation in this repository is excellent! Very clear and comprehensive.

eyestorefivem commented:
I'm impressed by the modularity of this codebase. Makes it very maintainable.

eyestoreoye commented:
Impressive project structure! The architecture really demonstrates solid design principles.

ayestoro commented:
The performance optimizations here are quite clever. Nice engineering work!

eyestorefivemstore commented:
Great implementation of this feature. I particularly like the approach to error handling.

fivemservereyestore commented:
The performance optimizations here are quite clever. Nice engineering work!

scriptforfivem commented:
I'm impressed by the modularity of this codebase. Makes it very maintainable.

moz-kathyreid commented:
## Refactored based on discussions with @moz-dfeller 

* There _already exist_ some sentences for `ug` in the `sentence_metadata` table in production
* So we need a way to INSERT a row, with `variant_id` into `sentence_metadata` if the `sentence_id` does not already exist in the table, but to UPDATE it if it already does. 
* We initially discussed using an UPDATE ON DUPLICATE KEY approach, _however_, `sentence_id` column in the `sentence_metadata` table is a _foreign key_ constraint, not a _primary key_ constraint so the UPDATE ON DUPLICATE KEY is not triggered. 

Instead, I approached this in two parts: 

1. Update the _existing_ records in `sentence_metadata` with the new variant
2. Insert _new_ records into `sentence_metadata` where they did not already exist

To test this, I: 

1. Deleted all existing rows for `ug` variants in `sentence_metadata`
2. Seeded the table with the following 10 rows which have null `variant_id`: 

```
INSERT INTO sentence_metadata (sentence_id, created_at)
VALUES ('fffb7a7b400a2677092451fb8dd4852915b833478bff75e59b0028807805d283', NOW())

INSERT INTO sentence_metadata (sentence_id, created_at)
VALUES ('fff9f1d40b87002c1e17f5d0cbf8579b72f438f1d9b88cf03a2ea346ea8e0a2c', NOW())

INSERT INTO sentence_metadata (sentence_id, created_at)
VALUES ('fff09b3c774316891ae9179c8350ca0dc83e9ddf5c30fc2df7dbe357225fa565', NOW())

INSERT INTO sentence_metadata (sentence_id, created_at)
VALUES ('ffeed1f099fc9ee842859f6240bc53a307bc3bd31c6bbe9bf271ab088ce7ac52', NOW())

INSERT INTO sentence_metadata (sentence_id, created_at)
VALUES ('ffed42e558076d9e017b174fe82e5618cf31719afdf511c9ca6fbb6d60ec9a78', NOW())

INSERT INTO sentence_metadata (sentence_id, created_at)
VALUES ('ffe98cab65a55efc87b2b5e4cac528e74dded9ca3939eed86aa0eadae0ebd9a5', NOW())

INSERT INTO sentence_metadata (sentence_id, created_at)
VALUES ('ffe01ec16e9a9d6553a3d0685142fea357f57ce97b53dbe699b0204dba3545ce', NOW())

INSERT INTO sentence_metadata (sentence_id, created_at)
VALUES ('ffde1c90813b233ea8d1a3a0760634fa4f350f207802fbc12dd18d0b3159ec89', NOW())

INSERT INTO sentence_metadata (sentence_id, created_at)
VALUES ('ffd5b2fdf9933c4f6107c70629a93e0d714437a2ccf8d85b7415af6cc7fe4bd5', NOW())

INSERT INTO sentence_metadata (sentence_id, created_at)
VALUES ('ffd5954b414f497871f305847b2cdda9527fda728fce787c1662372c3ffd580b', NOW())
```

3. Deleted _existing_ migration from `migrations` table
4. Ran `docker compose up` => check for successful migration run: 

```
common-voice  | [BE] [INFO] Processed migration 02505082230990-set-empty-ug-variants-to-arab
common-voice  | [BE] [INFO] Done
```

5. Check that all rows in DB for `ug` have expected `variant_id` in `sentence_metadata`: 

```
SELECT COUNT(*) AS CNT 
FROM sentence_metadata sm 
JOIN variants v on v.variant_token = 'ug-Arab'
WHERE v.id = sm.variant_id 
```

Expected value = 13,769 (13,759 new, 10 to update) = 13,769 in database

Validation checks passed, sending for review and merge.

moz-kathyreid commented:
Thanks @moz-dfeller - yeah this was a good exercise in MySQL - I haven't worked heavily with it for 10 years or so, excellent refresher. I really enjoy SQL so this was a fun challenge. 

_Excellent_ pickup on the filename - very precise, I like it. 

This is now renamed, sending back for review / merge.
